core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala

```diff
@@ -365,12 +387,46 @@ final class ShuffleBlockFetcherIterator(
     }
   }
 
+
+  /**
+   * Fetch the host-local blocks while we are fetching remote blocks. This is ok because
+   * `ManagedBuffer`'s memory is allocated lazily when we create the input stream, so all we
+   * track in-memory are the ManagedBuffer references themselves.
+   */
+  private[this] def fetchHostLocalBlocks() {
+    logDebug(s"Start fetching host-local blocks: ${hostLocalBlocks.mkString(", ")}")
+
+    val localDirsByExec =
+      blockManager.master.getHostLocalDirs(hostLocalBlocksByExecutor.keySet.toArray).localDirs
```

Comment:
the only reason I can see for ever turning this feature off is the extra roundtrip to the driver required here.  But instead of doing this once-per-task, could we add a cache on each executor for this data?  It shouldn't change much, but I guess you would need to limit the size as you could in theory have execs come and go a ton.

Reply:
I think having a cache in the executor is very good idea. First I will add them without any limit as the question regarding the limits is still pending.

Reply:
Actually I think the limit here is not needed at all as this cache will contain only the host local dirs of the executors running on this specific node where the requester executor is.